### PR TITLE
Migrate CI from Travis to GitHub actions

### LINKED ISSUE #707: Re-evaluate which CI runner to use
>> We may want to switch to Azure pipelines at some point in the not too distant future given the recent acquisition of Travis and some of the problems other projects have encountered.

@jdufresne mentioned:
> GitHub Actions might also be worth considering. It is fast and pleasant to work with. Integrating is only a matter of adding a new file.

Oh indeed, I didn't realize that was a job runner... I thought it was just a "trigger" thing that still required hooking up your own CI runner. I haven't looked into either one deeply, my comment above was just based on noticing a lot of popular projects moving to Azure.

This is probably better handled in a dedicated issue at some point where we can research and discuss more. For example, I know Azure allows 10 simultaneous pipelines, haven't checked what GA supports. (Although this project doesn't need 10, having more than one would be nice.)

### PR DESCRIPTION
**Pull request by jdufresne**
GitHub actions produces faster results than Travis, resulting in faster
feedback for contributors.

The linkcheck environment was not ported to GitHub actions. This
environment consistently fails with the error:

    429 Client Error: too many requests for url: https://github.com/…

This sometimes happens in Travis, but is more frequent now with the
increase in performance. See issue #771.

The "basepython" tox configuration option was dropped as GitHub actions
default to using the latest Python version which is now 3.9. This avoid
additional configuration in both actions and tox when a new Python is
released.

Fixes #707

### CODE REVIEW

File: .github/workflows/check.yml [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+      matrix:
+        tox-environment:
+          - docs
+          - examples
+          - lint
+
+    steps:
+      - uses: actions/checkout@v2
+
+      - name: Set up Python ${{ matrix.python-version }}
```
Comment (rbarrois) [inline .github/workflows/check.yml:22]: This does not look to be defined above?

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    steps:
+      - uses: actions/checkout@v2
+
+      - name: Set up Python ${{ matrix.python-version }}
+        uses: actions/setup-python@v2
+        with:
+          python-version: ${{ matrix.python-version }}
+
+      - name: Start MongoDB
+        uses: supercharge/mongodb-github-action@1.3.0
```
Comment (rbarrois) [inline .github/workflows/test.yml:34]: I am not familiar with the actions repository; how did you choose this action?

**Review by rbarrois — COMMENTED** [review]
Great idea, thanks — and quite fitting seeing Travis' latest pricing changes.

I've got a couple of questions though; moreover, we'll have to check setting this up in Github, and verify the pricing on Github's side.

File: .github/workflows/test.yml (hunk 2 above)
Comment (MRigal) [reply .github/workflows/test.yml:34]: It's the most commonly use as for now: https://github.com/marketplace?type=actions&query=mongodb

MongoDB Inc didn't feel the need to support official one until now.

It's not super great, but the risks are limited and it is the best option for now.

Maybe add a todo to check that in the future?

File: .github/workflows/check.yml (hunk 1 above)
Comment (MRigal) [reply .github/workflows/check.yml:22]: This is indeed not complete. I strongly recommend using https://github.com/ymyzk/tox-gh-actions

I let you check it @jdufresne , if required I could make PR for this PR

Comment (jdufresne) [reply .github/workflows/check.yml]: Thanks for catching this. Fixed in the latest revision.

File: .github/workflows/test.yml (hunk 2 above)
Comment (jdufresne) [reply .github/workflows/test.yml:34]: I chose this one as it seemed to be the most popular and common choice from my limited research.

If you have a preference for a different approach, please let me know and I'll take a look.

File: .github/workflows/check.yml (hunk 1 above)
Comment (jdufresne) [reply .github/workflows/check.yml]: @MRigal I took at look at tox-gh-actions and didn't find it to be an improvement, what in particular are you hoping to see?

To me, it spreads the configuration around. Before all the GH configuration is in the `workflows` directory. But with tox-gh-actions, it now also live in `tox.ini` or `setup.cfg`.

Additionally, different actions have different needs. For example, testing requires installing and running a MongoDB server. But running `flake8` does not require this. So I feel there is some advantage to keeping these configurations separate.

**Comment by jdufresne** [comment]
> we'll have to check setting this up in Github, and verify the pricing on Github's side.

This will turn on automatically once this lands on the `master` branch. Just the presence of these files is enough.

GitHub Actions usage is free for public repositories and self-hosted runners, so I don't think this is a concern. See here for more details:

https://docs.github.com/en/free-pro-team@latest/actions/reference/usage-limits-billing-and-administration

File: .travis.yml [hunk 3]
```diff
@@ -1,38 +0,0 @@
...
-      env: TOXENV=examples
-    - python: "3.9"
-      env: TOXENV=lint
-
-services:
-  - mongodb
-
-notifications:
-  email: false
-  irc: "irc.freenode.org#factory_boy"
```
Comment (rbarrois) [inline .travis.yml:38]: I believe we're losing that part though :(

File: .github/workflows/test.yml (hunk 2 above)
Comment (rbarrois) [reply .github/workflows/test.yml:38]: Based on [this issue](https://github.com/actions/example-services/issues/8#issuecomment-559484139), I believe it should be possible with:

```yaml
services:
  mongodb:
    image: mongo:latest
    ports:
      - 27017:27017
```

(And that's basically what I use locally for development, too :wink:)

Comment (jdufresne) [reply .github/workflows/test.yml]: Nice, this worked! Thanks for pointing this out.

I've added this in the latest version. You can see it working here:

https://github.com/jdufresne/factory_boy/actions/runs/373502935

File: .travis.yml (hunk 3 above)
Comment (jdufresne) [reply .travis.yml:38]: That's true. Do you consider this a blocker?

There are some IRC actions available from GH marketplace:

https://github.com/marketplace?query=irc

However, they aren't very compelling to me. For one, they seem to require a lot of customization per message and situation while the Travis version automatically handled different scenarios. If you think one of these should be included, let me know and I can experiment with them.

I suppose the alternative would be to use an in house custom action or command line tool.

**Review by rbarrois — APPROVED** [review]
@jdufresne Let's go and merge it as is, I'll find a way to fix IRC notifications later if needed :)

**Pull request merged**